README.md
# waterData

See the following report for a complete description of the project: Ryberg, K.R., and Vecchia, A.V., 2012, waterData—An R package for retrieval, analysis, and anomaly calculation of daily hydrologic time series data, version 1.0: U.S. Geological Survey Open-File Report 2012–1168, 8 p. https://pubs.usgs.gov/of/2012/1168/.


# Disclaimer
----------
This software is in the public domain because it contains materials that originally came from the U.S. Geological Survey (USGS), an agency of the United States Department of Interior. For more information, see the official USGS copyright policy at [https://www.usgs.gov/visual-id/credit_usgs.html#copyright](https://www.usgs.gov/visual-id/credit_usgs.html#copyright)

Although this software program has been used by the USGS, no warranty, expressed or implied, is made by the USGS or the U.S. Government as to the accuracy and functioning of the program and related program material nor shall the fact of distribution constitute any such warranty, and no responsibility is assumed by the USGS in connection therewith.

This software is provided "AS IS."

 [
   ![CC0](http://i.creativecommons.org/p/zero/1.0/88x31.png)
 ](http://creativecommons.org/publicdomain/zero/1.0/)


NEWS
Version 1.0.8
*************
Bug fix in importDVs.

Version 1.0.7
*************
Minor bug fixes.

Version 1.0.6
*************
Removed XML from imports as it is no longer used.

Version 1.0.5
*************
Updated all USGS urls to https.

Version 1.0.4
*************
Fixed code that was causing an error in the Vignette.
Changed depends list to imports for additional packages used by 
some of the functions.
Added a more descriptive error message for long-term anomaly 
calculation attempt on datasets that are too short.

Version 1.0.3
*************
Fixed incomplete plot title in plotAnoms when which parameter equals 4.
Added NEWS file.
Migrated vignette to 'vignettes' directory.

Version 1.0.2
*************
Minor bug fixes.

Version 1.0.1
*************
Minor bug fixes.

Version 1.0.0
*************
This is the first publicly available version of waterData.


fillMiss.R
#' Function to fill in missing time series data.
#' 
#' This function will check the percent of missing values and the size
#' of the largest missing block of data.  By default, if less than 40 percent 
#' of the data are missing and the largest block is less than 30 days, the 
#' data will be filled-in by using a structural time series, \link{StructTS} 
#' from the base stats  package in R (R Development Core Team, 2012).  The 
#' fitted structural time series is then smoothed via a state-space model, 
#' \link{tsSmooth} from the base stats package in R.
#' 
#' @name fillMiss
#' @title Fill-in missing hydrological values
#' @param dataset is a data frame in the format of the data frame returned by
#' \link{importDVs}, with missing values indicated by NA.
#' @param block is the size of the largest block of missing data that the 
#' function will fill-in.
#' @param pmiss is the maximum amount of the missing data that can be missing
#' in the dataset for fill-in procedure to be performed.
#' @param model is the type of structural time series model, see 
#' \link{StructTS}.  The default value is trend.  If level is used, the results 
#' of \link{fillMiss}, which by default applies a fixed-interval smoothing to 
#' the time series, \link{tsSmooth}, will be very close to linear interpolation. 
#' @param smooth  a logical that indicates whether or not to apply 
#' \link{tsSmooth} to the structured time series.
#' @param ... further arguments to be passed to plotting method (see \link{par}).
#' @return a data frame with NAs in the "val" column replaced by 
#' estimated values and a plot showing observed and estimated data.  If there 
#' are too many missing values, based on default or user defined limits, the
#' unaltered dataset is returned as well as a message, such as "Error 
#' in fillMiss(misQ05054000) : Too much missing data.  Cannot fill in missing 
#' values."
#' @note Many methods have been suggested for estimating missing hydrological 
#' data.  However, experiments showed that the functions in the base stats 
#' package  worked very well if the blocks of missing data were not long.  
#' Users with larger blocks of missing data may want to explore other methods 
#' including using nearby gages to estimate missing values at a streamgage.  
#' Additional methods for filling in missing hydrological data are summarized 
#' in Beauchamp and others (1989) and Elshorbagy and others (2000).
#' 
#' To indicate which values have been replaced, the qualcode field is
#' populated with 'fM' for those values that were estimated using the
#' fillMiss function.
#' @export
#' @seealso \link{StructTS}, \link{tsSmooth}, \link{cleanUp}
#' @format The returned data frame has the following columns: \cr
#' \tabular{lll}{
#' Name \tab Type \tab Description \cr 
#' staid \tab factor \tab USGS station identification number \cr
#' val \tab numeric \tab The value of the hydrologic variable \cr
#' dates \tab Date \tab Date of daily value \cr
#' qualcode \tab factor \tab Qualification code
#' }
#' @examples
#' data(exampleWaterData)
#' my.newdata <- fillMiss(misQ05054000, block=30, pmiss=50, log="y")
#' summary(misQ05054000)
#' summary(my.newdata)
#' # ph example
#' pH05082500<-importDVs("05082500", code="00400", stat="00008", 
#' sdate="2000-01-01", edate="2011-12-31")
#' plotParam(pH05082500)
#' pHfilled<-fillMiss(pH05082500, block=45, ylim=c(7.5,9), yaxs="i")
#' @keywords NA hplot datagen ts smooth
#' @references
#' Beauchamp, J.J., 1989, Comparison of regression and time-series methods for
#' synthesizing missing streamflow records: Water Resources Bulletin, v. 25, 
#' no. 5, p. 961--975.
#'
#' Elshorbagy, A.A., Panu, U.S., Simonovic, S.P., 2000, Group-based
#' estimation of missing hydrological data---I. Approach and general
#' methodology: Hydrological Sciences Journal, v. 45, no. 6, p. 849--866.
#' 
#' R Development Core Team, 2012, R---A language and environment for statistical
#' computing: Vienna, Austria, R Foundation for Statistical Computing, [ISBN
#' 3-900051-07-0].  (Also available at \url{https://www.r-project.org}.)
fillMiss <- function(dataset, block=30, pmiss=40, model="trend", 
                     smooth=TRUE, ... ) {
  pck<-is.na(dataset$val)
  percent<-0
  max.mis<-0
  if (sum(pck) > 0) {
    percent<-round((sum(pck)/length(dataset$val))*100, digits=2)
    my.message<-paste("There are ", sum(pck), 
                      "missing values for", dataset$staid[1], "or", percent, 
                      "percent of the data.", sep=" ")
    message(my.message)
    rles<-rle(is.na(dataset$val))
    max.mis<-max(rles$lengths[rles$values])
    my.message2<-paste("The maximum block of missing", dataset$staid[1], 
                       "data is", max.mis, "days long.", sep=" ")
    message(my.message2)
  } else { noMis<-paste("No missing values for", dataset$staid[1], sep=" ")
           message(noMis)
  }
  if ( percent >= pmiss | max.mis >= block ) {
    tooMuch <- paste("Too much missing data for", dataset$staid[1], 
                     "Cannot fill in missing values.", sep=" ")
    message (tooMuch)
  } else {
    my.series<-window(dataset$val)
    my.struct<-StructTS(my.series, type=model)
    if ( smooth ) fit<-tsSmooth(my.struct) else fit<-fitted(my.struct)
    #par(las=1, tck=0.01)
    plot(my.series, typ="l", lwd=4, xlab="Observation", 
         ylab="Observed and estimated times series", ...)
    lines(fit[,1],col="green")
    leg.txt<-c("Observed values", "New time series")
    legend("topleft", leg.txt, col=c("black","green"), lwd=c(4,1), bty="n",
           ncol=2, cex=0.8)
    dataset$val[pck] <- fit[pck,1]
    lng <- length(grep("fM", levels(dataset$qualcode)))
    if ( lng < 1 ) {
      dataset$qualcode <- factor(dataset$qualcode, levels=c(levels(dataset$qualcode), 'fM')) 
    }
    dataset$qualcode[pck] <- "fM"
    if (!is.null(attributes(dataset)$stat) & 
      !is.null(attributes(dataset)$code)) {
      subtext <- paste("Parameter code", attributes(dataset)$code, 
                       " Statistics code", attributes(dataset)$stat, 
                       sep=" ")
      title(sub=subtext)
    }
    fillMess <- paste("Filled in", sum(pck), "values for", dataset$staid[1],
                      sep=" ")
    message (fillMess)
  }
  dataset
}


plotAnoms.R
#' Function to plot hydrologic time series and anomalies
#'
#' @name plotAnoms
#' @title Plots streamflow anomalies
#' @param data is the anomaly list from the function 
#' \link{compAnom}.
#' @param ... further arguments to be passed to plotting method (see \link{par}).
#' @return a plot.
#' @keywords hplot ts multivariate
#' @export
#' @examples
#' q05054000.85 <- importDVs("05054000", sdate="1985-01-01", edate="2010-09-30")
#' anoms05054000 <- compAnom(q05054000.85, which=1)
#' plotAnoms(anoms05054000)
plotAnoms<- function(data, ...) {
  if ( ! is.data.frame(data) ) {
    if ( length(data) == 4 ) {
      par(mfrow=c(2,2), cex.lab=.9, las=1, tcl=0.5, xaxs="r", yaxs="r", 
          mar=c(4, 5, 0, 1) + 0.1,mgp=c(4, 1, 0),oma=c(0,0,1,0)+.1)
      plot(data[[1]]$dates, data[[1]]$val, log="y", type="l", xlab="",
           ylab="Streamflow, cubic feet per second", cex.lab=0.6, ...)
      plot(data[[1]]$dates, data[[1]]$ltfa, type="l", xlab="", 
           ylab="Long-term flow anomaly", cex.lab=0.6, ...)
      ltfa.sd <- round(sd(data[[1]]$ltfa, na.rm=TRUE), digits=3)
      mtext(paste("Standard deviation is ", ltfa.sd, sep=""),side=3, line=-1, 
            cex=0.5)
      plot(data[[1]]$dates, data[[1]]$mtfa, type="l", xlab="", 
           ylab="Medium-term flow anomaly", cex.lab=0.6, ...)
      mtfa.sd<-round(sd(data[[1]]$mtfa, na.rm=TRUE), digits=3)
      mtext(paste("Standard deviation is ", mtfa.sd, sep=""), side=3, line=-1, 
            cex=0.5)
      plot(data[[1]]$dates, data[[1]]$stfa, type="l", xlab="", 
           ylab="Short-term flow anomaly", cex.lab=0.6, ...)
      stfa.sd<-round(sd(data[[1]]$stfa, na.rm=TRUE), digits=3)
      mtext(paste("Standard deviation is ", stfa.sd, sep=""), side=3, line=-1, 
            cex=0.5)
      plttxt<-paste("Site", as.character(data[[1]]$staid[1]), "LTFA is", data[[2]],
                    "days", "MTFA is", data[[3]], "days", "STFA is", data[[4]],
                    sep=" ")
      mtext(plttxt, outer=TRUE, side=3, cex=.8)
    }
    else if ( length(data) == 3 ) {
      par(mfrow=c(2,2), cex.lab=.9, las=1, tcl=0.5, xaxs="r", yaxs="r", 
      mar=c(4, 5, 0, 1) + 0.1,mgp=c(4, 1, 0),oma=c(0,0,1,0)+.1)
      plot(data[[1]]$dates, data[[1]]$val, log="y", type="l", xlab="",
           ylab="Streamflow, cubic feet per second", cex.lab=0.6, ...)
      plot(data[[1]]$dates, data[[1]]$mtfa, type="l", xlab="", 
           ylab="Medium-term flow anomaly", cex.lab=0.6, ...)
      mtfa.sd<-round(sd(data[[1]]$mtfa, na.rm=TRUE), digits=3)
      mtext(paste("Standard deviation is ", mtfa.sd, sep=""), side=3, line=-1, 
            cex=0.5)
      plot(data[[1]]$dates, data[[1]]$stfa, type="l", xlab="", 
           ylab="Short-term flow anomaly", cex.lab=0.6, ...)
      stfa.sd<-round(sd(data[[1]]$stfa, na.rm=TRUE), digits=3)
      mtext(paste("Standard deviation is ", stfa.sd, sep=""), side=3, line=-1, 
            cex=0.5)
      plttxt<-paste("Site", as.character(data[[1]]$staid[1]),
                    "days", "MTFA is", data[[2]], "days", "STFA is", data[[3]],
                    sep=" ")
      mtext(plttxt, outer=TRUE, side=3, cex=.8)
    }
    else if ( length(data) == 6 ) {
      par(mfrow=c(3,2), cex.lab=.9, las=1, tcl=0.5, xaxs="r", yaxs="r", 
          mar=c(4, 5, 0, 1) + 0.1,mgp=c(4, 1, 0),oma=c(0,0,1,0)+.1)
      plot(data[[1]]$dates, data[[1]]$val, log="y", type="l", xlab="",
           ylab="Streamflow, cubic feet per second", cex.lab=0.6, ...)
      plot(data[[1]]$dates, data[[1]]$tenyranom, type="l", xlab="", 
           ylab="Ten-year flow anomaly", cex.lab=0.6, ...)
      tfa.sd <- round(sd(data[[1]]$tenyranom, na.rm=TRUE), digits=3)
      mtext(paste("Standard deviation is ", tfa.sd, sep=""),side=3, line=-1, 
            cex=0.5)
      plot(data[[1]]$dates, data[[1]]$fiveyranom, type="l", xlab="", 
           ylab="Five-year flow anomaly", cex.lab=0.6, ...)
      ffa.sd <- round(sd(data[[1]]$fiveyranom, na.rm=TRUE), digits=3)
      mtext(paste("Standard deviation is ", ffa.sd, sep=""),side=3, line=-1, 
            cex=0.5)
      plot(data[[1]]$dates, data[[1]]$annualanom, type="l", xlab="", 
         ylab="One-year flow anomaly", cex.lab=0.6, ...)
      ltfa.sd <- round(sd(data[[1]]$annualanom, na.rm=TRUE), digits=3)
      mtext(paste("Standard deviation is ", ltfa.sd, sep=""),side=3, line=-1, 
            cex=0.5)
      plot(data[[1]]$dates, data[[1]]$seasanom, type="l", xlab="", 
           ylab="Seasonal flow anomaly", cex.lab=0.6, ...)
      mtfa.sd<-round(sd(data[[1]]$seasanom, na.rm=TRUE), digits=3)
      mtext(paste("Standard deviation is ", mtfa.sd, sep=""), side=3, line=-1, 
            cex=0.5)
      plot(data[[1]]$dates, data[[1]]$dailyanom, type="l", xlab="", 
           ylab="Daily flow anomaly", cex.lab=0.6, ...)
      stfa.sd<-round(sd(data[[1]]$dailyanom, na.rm=TRUE), digits=3)
      mtext(paste("Standard deviation is ", stfa.sd, sep=""), side=3, line=-1, 
            cex=0.5)
      plttxt<-paste("Site", as.character(data[[1]]$staid[1]), "Ten-year, 
                    five-year, one-year, seasonal, and daily anomalies",
                    sep=" ")
      mtext(plttxt, outer=TRUE, side=3, cex=.8)
    }
    else {
      stop("Data supplied must be a list of length 3, 4, or 6 generated by the 
           compQAnom function.")
    }
  }
  else if (is.data.frame(data)) {
    if ( length(data) == 6 ) {
      par(mfrow=c(2,2), cex.lab=.9, las=1, tcl=0.5, xaxs="r", yaxs="r", 
          mar=c(4, 5, 0, 1) + 0.1,mgp=c(4, 1, 0),oma=c(0,0,1,0)+.1)
      plot(data$dates, data$val, log="y", type="l", xlab="",
           ylab="Streamflow, cubic feet per second", cex.lab=0.6, ...)
      plot(data$dates, data$ltfa, type="l", xlab="", 
           ylab="Long-term flow anomaly", cex.lab=0.6, ...)
      ltfa.sd <- round(sd(data$ltfa, na.rm=TRUE), digits=3)
      mtext(paste("Standard deviation is ", ltfa.sd, sep=""),side=3, line=-1, 
            cex=0.5)
      plot(data$dates, data$mtfa, type="l", xlab="", 
           ylab="Medium-term flow anomaly", cex.lab=0.6, ...)
      mtfa.sd<-round(sd(data$mtfa, na.rm=TRUE), digits=3)
      mtext(paste("Standard deviation is ", mtfa.sd, sep=""), side=3, line=-1,
            cex=0.5)
      plot(data$dates, data$stfa, type="l", xlab="", 
           ylab="Short-term flow anomaly", cex.lab=0.6, ...)
      stfa.sd<-round(sd(data$stfa, na.rm=TRUE), digits=3)
      mtext(paste("Standard deviation is ", stfa.sd, sep=""), side=3, line=-1,
            cex=0.5)
      plttxt<-paste("Site", as.character(data$staid[1]), sep=" ")
      mtext(plttxt, outer=TRUE, side=3, cex=.8)
    }
    else if ( length(data) == 5 ) {
      par(mfrow=c(2,2), cex.lab=.9, las=1, tcl=0.5, xaxs="r", yaxs="r", 
          mar=c(4, 5, 0, 1) + 0.1,mgp=c(4, 1, 0),oma=c(0,0,1,0)+.1)
      plot(data$dates, data$val, log="y", type="l", xlab="",
           ylab="Streamflow, cubic feet per second", cex.lab=0.6, ...)
      plot(data$dates, data$mtfa, type="l", xlab="", 
           ylab="Medium-term flow anomaly", cex.lab=0.6, ...)
      mtfa.sd<-round(sd(data$mtfa, na.rm=TRUE), digits=3)
      mtext(paste("Standard deviation is ", mtfa.sd, sep=""), side=3, line=-1, 
            cex=0.5)
      plot(data$dates, data$stfa, type="l", xlab="", 
           ylab="Short-term flow anomaly", cex.lab=0.6, ...)
      stfa.sd<-round(sd(data$stfa, na.rm=TRUE), digits=3)
      mtext(paste("Standard deviation is ", stfa.sd, sep=""), side=3, line=-1, 
            cex=0.5)
      plttxt<-paste("Site", as.character(data$staid[1]), sep=" ")
      mtext(plttxt, outer=TRUE, side=3, cex=.8)
    }
    else if ( length(data) == 8 ) {
      par(mfrow=c(3,2), cex.lab=.9, las=1, tcl=0.5, xaxs="r", yaxs="r", 
          mar=c(4, 5, 0, 1) + 0.1,mgp=c(4, 1, 0),oma=c(0,0,1,0)+.1)
      plot(data$dates, data$val, log="y", type="l", xlab="",
           ylab="Streamflow, cubic feet per second", cex.lab=0.6, ...)
      plot(data$dates, data$tenyranom, type="l", xlab="", 
           ylab="Ten-year flow anomaly", cex.lab=0.6, ...)
      tfa.sd <- round(sd(data$tenyranom, na.rm=TRUE), digits=3)
      mtext(paste("Standard deviation is ", tfa.sd, sep=""),side=3, line=-1, 
            cex=0.5)
      plot(data$dates, data$fiveyranom, type="l", xlab="", 
           ylab="Five-year flow anomaly", cex.lab=0.6, ...)
      ffa.sd <- round(sd(data$fiveyranom, na.rm=TRUE), digits=3)
      mtext(paste("Standard deviation is ", ffa.sd, sep=""),side=3, line=-1, 
            cex=0.5)
      plot(data$dates, data$annualanom, type="l", xlab="", 
           ylab="One-year flow anomaly", cex.lab=0.6, ...)
      ltfa.sd <- round(sd(data$annualanom, na.rm=TRUE), digits=3)
      mtext(paste("Standard deviation is ", ltfa.sd, sep=""),side=3, line=-1, 
            cex=0.5)
      plot(data$dates, data$seasanom, type="l", xlab="", 
           ylab="Seasonal flow anomaly", cex.lab=0.6, ...)
      mtfa.sd<-round(sd(data$seasanom, na.rm=TRUE), digits=3)
      mtext(paste("Standard deviation is ", mtfa.sd, sep=""), side=3, line=-1, 
            cex=0.5)
      plot(data$dates, data$dailyanom, type="l", xlab="", 
           ylab="Daily flow anomaly", cex.lab=0.6, ...)
      stfa.sd<-round(sd(data$dailyanom, na.rm=TRUE), digits=3)
      mtext(paste("Standard deviation is ", stfa.sd, sep=""), side=3, line=-1, 
            cex=0.5)
      plttxt<-paste("Site", as.character(data$staid[1]), "Ten-year, 
                  five-year, one-year, seasonal, and daily anomalies",
                  sep=" ")
      mtext(plttxt, outer=TRUE, side=3, cex=.8)
    }
    else {
      stop("Data supplied must be a data frame with 5, 6, or 8 columns.")
    }  
  }
  else {
    stop("Data must be a list or dataframe.")
  }
}

waterData-package.R
#' Retrieval, analysis, and anomaly calculation of daily 
#' hydrologic time deries data.
#'
#' This package imports U.S. Geological Survey (USGS) daily hydrologic data 
#' from USGS web services, plots the data, addresses some common data problems, 
#' and calculates and plots anomalies.  For a description of anomalies see 
#' Vecchia (2003), and for examples of the application of streamflow anomalies 
#' in trend analysis of nutrients, pesticides and surface water, see Alexander 
#' and Smith (2006), Ryberg and Vecchia (2006), Ryberg and others (2010), 
#' Sullivan and others (2009), Vecchia (2005), and Vecchia and others (2008).
#'
#' \tabular{ll}{
#' Package: \tab waterData\cr
#' Type: \tab Package\cr
#' Version: \tab 1.0.8\cr
#' Date: \tab 2017-04-28\cr
#' License: \tab Unlimited | file LICENSE \cr
#' LazyLoad: \tab yes\cr
#' }
#'
#' @name waterData-package
#' @aliases waterData
#' @docType package
#' @title Hydrologic Data Retrieval and Analysis and Anomaly Calculation
#' @author Karen R. Ryberg \email{kryberg@@usgs.gov} and 
#' Aldo V. Vecchia \email{avecchia@@usgs.gov}
#' @keywords package
#' @references 
#' Alexander, R.B. and Smith, R.A., 2006, Trends in the nutrient enrichment of 
#' U.S. rivers during the late 20th century and their relation to changes in
#' probable stream trophic conditions: Limnology and Oceanography, v. 51, no.
#' 1, Part 2: Eutrophication of Freshwater and Marine Ecosystems, p. 
#' 639--654., accessed August 1, 2012, at \url{http://www.jstor.org/stable/4499617}.
#'
#' Ryberg, K.R. and Vecchia, A.V., 2006, Water-quality trend analysis and 
#' sampling design for the Devils Lake Basin, North Dakota, January 1965 
#' through September 2003: U.S. Geological Survey Scientific Investigations 
#' Report 2006--5238, 64 p., accessed August 1, 2012, at
#'  \url{http://pubs.usgs.gov/sir/2006/5238/}.
#'
#' Ryberg, K.R. and Vecchia, A.V., 2012, waterData---An R package for retrieval, 
#' analysis, and anomaly calculation of daily hydrologic time series data, 
#' version 1.0: U.S. Geological Survey Open-File Report 2012--1168, 8 p. 
#' (Also available at \url{http://pubs.usgs.gov/of/2012/1168/}.)
#' 
#' Ryberg, K.R., Vecchia, A.V., Martin, J.D., Gilliom, R.J., 2010, Trends in 
#' pesticide concentrations in urban streams in the United States, 1992--2008: 
#' U.S. Geological Survey Scientific Investigations Report 2010-5139, 101 p., 
#' accessed August 1, 2012, at \url{http://pubs.usgs.gov/sir/2010/5139/}.
#'
#' Sullivan, D.J., Vecchia, A.V., Lorenz, D.L., Gilliom, R.J., Martin, J.D., 
#' 2009, Trends in pesticide concentrations in corn-belt streams, 1996--2006: 
#' U.S. Geological Survey Scientific Investigations Report 2009-5132, 75 p., 
#' accessed August 1, 2012, at \url{http://pubs.usgs.gov/sir/2009/5132/}.
#'
#' Vecchia, A.V., 2003, Relation between climate variability and stream water 
#' quality in the continental United States, Hydrological Science and 
#' Technology, v. 19 no. 1, 77--98.
#'
#' Vecchia, A.V., 2003, Water-quality trend analysis and sampling design for 
#' streams in North Dakota, 1971--2000: U.S. Geological Survey Scientific 
#' Investigations Report 2003--4094, 73 p., accessed August 1, 2012, at  
#' \url{http://nd.water.usgs.gov/pubs/wri/wri034094/index.html}.
#'
#' Vecchia, A.V., 2005, Water-quality trend analysis and sampling design for 
#' streams in the Red River of the North Basin, Minnesota, North Dakota, and 
#' South Dakota, 1970--2001: U.S. Geological Survey Scientific Investigations 
#' Report 2005--5224, 54 p., accessed August 1, 2012, at 
#' \url{http://pubs.usgs.gov/sir/2005/5224/}.
NULL


compAnom.R
#' Function to calculate short-, medium-, and long-term hydrologic anomalies
#'
#' This function was written with streamflow data in mind because streamflow 
#' is the most commonly used exogenous variable for trend models for water 
#' quality;  however, the function is generic so that users may experiment with 
#' anomalies from other daily hydrologic data.  Examples of the inclusion of 
#' streamflow anomalies in trend analysis of nutrients, pesticides and 
#' surface water can be found in Alexander and Smith (2006), Ryberg and Vecchia 
#' (2006), Ryberg and others (2010), Sullivan and others (2009), Vecchia
#' (2003), Vecchia (2005), and Vecchia and others (2008).
#'
#' @name compAnom
#' @title Calculates  anomalies
#' @param dataset is the daily hydrologic data returned from \link{importDVs} 
#' or data otherwise obtained and in the same format as that produced by 
#' \link{importDVs}.
#' @param which indicates which set of anomalies; 1 calculates the 1-year, 
#' 30-day, and 1-day anomalies; 2 calculates the 100-day, 10-day, and 1-day 
#' anomalies; 3 calculates the 30-day and 1-day anomalies; and 4 calculates 
#' the 10-year, 5-year, 1-year, one-quarter-year (seasonal), and 1-day 
#' anomalies.
#' @keywords ts multivariate
#' @return A list.  In the cases of "which" equal to 1 or 2, the 
#' first element of the list is a data frame containing the station 
#' identification number, dates, streamflow, and long-term, mid-term, and 
#' short-term anomalies.  The next three elements of the list are the length 
#' in days of the long-term, mid-term, and short-term streamflow anomalies.  
#' In the case of "which" equal to 3, the first element of the list is a data 
#' frame containing the station identification number, dates, streamflow, and 
#' mid-term and short-term anomalies.  The next two elements of the list are 
#' the length in days of the mid-term and short-term streamflow anomalies.  
#' In the case of "which" equal to 4, the first element of the list is a data 
#' frame containing the station identification number, dates, streamflow, and 
#' 10-year, 5-year, annual, seasonal, and daily streamflow anomalies.  The 
#' next five elements of the list are the length in days of the 10-year, 
#' 5-year, annual, seasonal, and daily streamflow anomalies.  
#' @export
#' @examples
#' q05054000.85 <- importDVs("05054000", sdate="1985-10-01", edate="2010-09-30")
#' anoms05054000.1 <- compAnom(q05054000.85, which=1)
#' anoms05054000.2 <- compAnom(q05054000.85, which=2)
#' anoms05054000.3 <- compAnom(q05054000.85, which=3)
#' anoms05054000.4 <- compAnom(q05054000.85, which=4)
#' @references
#' Alexander, R.B. and Smith, R.A., 2006, Trends in the nutrient enrichment of 
#' U.S. rivers during the late 20th century and their relation to changes in
#' probable stream trophic conditions: Limnology and Oceanography, v. 51, no.
#' 1, Part 2: Eutrophication of Freshwater and Marine Ecosystems, p. 
#' 639--654, accessed August 1, 2012, at 
#' \url{http://www.jstor.org/stable/4499617}.
#' 
#' Ryberg, K.R. and Vecchia, A.V., 2006, Water-quality trend analysis and 
#' sampling design for the Devils Lake Basin, North Dakota, January 1965 
#' through September 2003: U.S. Geological Survey Scientific Investigations 
#' Report 2006--5238, 64 p., accessed August 1, 2012, at 
#' \url{http://pubs.usgs.gov/sir/2006/5238/}.
#'
#' Ryberg, K.R. and Vecchia, A.V., 2012, waterData---An R packge for retrieval, 
#' analysis, and anomaly calculation of daily hydrologic time series data, 
#' version 1.0: U.S. Geological Survey Open-File Report 2012--1168, 8 p.  
#' (Also available at \url{http://pubs.usgs.gov/of/2012/1168/}.)
#' 
#' Ryberg, K.R., Vecchia, A.V., Martin, J.D., Gilliom, R.J., 2010, Trends in 
#' pesticide concentrations in urban streams in the United States, 1992-2008: 
#' U.S. Geological Survey Scientific Investigations Report 2010--5139, 101 p. 
#' (Also available at \url{http://pubs.usgs.gov/sir/2010/5139/}.)
#'
#' Sullivan, D.J., Vecchia, A.V., Lorenz, D.L., Gilliom, R.J., Martin, J.D., 
#' 2009, Trends in pesticide concentrations in corn-belt streams, 1996--2006: 
#' U.S. Geological Survey Scientific Investigations Report 2009--5132, 75 p., 
#' accessed Ausugst , 2012, at \url{http://pubs.usgs.gov/sir/2009/5132/}.
#'
#' Vecchia, A.V., 2003, Relation between climate variability and stream water 
#' quality in the continental United States, Hydrological Science and 
#' Technology, v. 19, no. 1, 77--98.
#'
#' Vecchia, A.V., 2003, Water-quality trend analysis and sampling design for 
#' streams in North Dakota, 1971--2000: U.S. Geological Survey Scientific 
#' Investigations Report 2003--4094, 73 p., accessed August 1, 2012, at 
#' \url{http://nd.water.usgs.gov/pubs/wri/wri034094/index.html}.
#'
#' Vecchia, A.V., 2005, Water-quality trend analysis and sampling design for 
#' streams in the Red River of the North Basin, Minnesota, North Dakota, and 
#' South Dakota, 1970--2001: U.S. Geological Survey Scientific Investigations 
#' Report 2005--5224, 54 p. accessed August 1, 2012, at 
#' \url{http://pubs.usgs.gov/sir/2005/5224/}.
compAnom <- function(dataset,which=1) {
  if (which==1) {
    lt<-365
    mt<-30
    st<-1
    n1 <- length(dataset[,1])
    qdf <- dataset[,c("staid","dates","val")]
    qlog10 <- log10(dataset$val)
    xmean <- mean(qlog10,na.rm=TRUE)  
    xlt <- rep(NA,n1)
    for (i in (lt:n1)) { 
      xlt[i] <- mean(qlog10[(i - (lt - 1)):i],na.rm=FALSE)
    }
    ltfa <- xlt-xmean
    xmt <- rep(NA,n1)
    for (i in (mt:n1)) { 
      xmt[i] <- mean(qlog10[(i - (mt - 1)):i],na.rm=FALSE)
    }
    mtfa <- xmt-xlt
    # short term anomaly is based on 1 day
    stfa <- qlog10 - xmt
  
    qdf <- cbind(qdf,ltfa,mtfa,stfa)
    list(qdf, lt,mt,st)
  }
  else if (which==2) {
    lt<-100
    mt<-10
    st<-1
    n1 <- length(dataset[,1])
    qdf <- dataset[,c("staid","dates","val")]
    qlog10 <- log10(dataset$val)
    xmean <- mean(qlog10,na.rm=TRUE)  
    xlt <- rep(NA,n1)
    for (i in (lt:n1)) { 
      xlt[i] <- mean(qlog10[(i - (lt - 1)):i],na.rm=FALSE)
    }
    ltfa <- xlt-xmean
    xmt <- rep(NA,n1)
    for (i in (mt:n1)) { 
      xmt[i] <- mean(qlog10[(i - (mt - 1)):i],na.rm=FALSE)
    }
    mtfa <- xmt-xlt
    # short term anomaly is based on 1 day
    stfa <- qlog10 - xmt
    
    qdf <- cbind(qdf,ltfa,mtfa,stfa)
    list(qdf,lt,mt,st)
  }
  else if (which==3) {
    st <- 1
    mt <- 30
    n1 <- length(dataset[,1])
    qdf <- dataset[,c("staid","dates","val")]
    qlog10 <- log10(dataset$val)
    xmean <- mean(qlog10,na.rm=TRUE)  
    xmt <- rep(NA,n1)
    for (i in (mt:n1)) { 
      xmt[i] <- mean(qlog10[(i - (mt - 1)):i],na.rm=FALSE)
    }
    mtfa <- xmt-xmean
    # short term anomaly is based on 1 day
    stfa <- qlog10 - xmt
  
    qdf <- cbind(qdf,mtfa,stfa)
    list(qdf, mt, st)
  }
  else if (which==4) {
    ta<-10*365
    fa<-5*365
    lt<-365
    mt<-90
    st<-1
    n1 <- length(dataset[,1])
    qdf <- dataset[,c("staid","dates","val")]
    qlog10 <- log10(dataset$val)
    xmean <- mean(qlog10,na.rm=TRUE)
    xta <- rep(NA,n1)
    for (i in (ta:n1)) { 
      xta[i] <- mean(qlog10[(i - (ta - 1)):i],na.rm=FALSE)
    }
    tfa <- xta-xmean
    
    xfa <- rep(NA,n1)
    for (i in (fa:n1)) { 
      xfa[i] <- mean(qlog10[(i - (fa - 1)):i],na.rm=FALSE)
    }    
    ffa <- xfa-xta
    
    xlt <- rep(NA,n1)
    for (i in (lt:n1)) { 
      xlt[i] <- mean(qlog10[(i - (lt - 1)):i], na.rm=FALSE)
    }    
    ltfa <- xlt-xfa
    
    xmt <- rep(NA,n1)
    for (i in (mt:n1)) { 
      xmt[i] <- mean(qlog10[(i - (mt - 1)):i],na.rm=FALSE)
    }
    mtfa <- xmt-xlt
    
    # short term anomaly is based on 1 day
    stfa <- qlog10 - xmt
  
    qdf <- cbind(qdf,tfa,ffa,ltfa,mtfa,stfa)
    dimnames(qdf)[[2]] <- c("staid", "dates", "val", "tenyranom", "fiveyranom",
                            "annualanom","seasanom","dailyanom")
    list(qdf, ta, fa, lt, mt, st)
  }
  else {
    stop("which must be a numeric value 1, 2, 3, or 4")
  }
}
